apt-hook/20apt-esm-hook.conf

```diff
@@ -1,7 +1,7 @@
 APT::Update::Pre-Invoke {
-	"[ ! -e /run/systemd/system ] || [ $(id -u) -ne 0 ] || systemctl start --no-block apt-news.service esm-cache.service || true";
+	"[ ! -e /run/systemd/system ] || [ $(id -u) -ne 0 ] || systemctl start --no-block apt-news.service esm-cache.service >/dev/null 2>&1 || true";
 };
 
 binary::apt::AptCli::Hooks::Upgrade {
-	"[ ! -f /usr/lib/ubuntu-advantage/apt-esm-json-hook ] || /usr/lib/ubuntu-advantage/apt-esm-json-hook || true";
+	"[ ! -f /usr/lib/ubuntu-advantage/apt-esm-json-hook ] || /usr/lib/ubuntu-advantage/apt-esm-json-hook 2> /var/lib/ubuntu-advantage/apt-upgrade-hook-err.txt || true";
```

Comment:
I'll take a closer look at this on Monday. The questions that come to mind at the moment are (and I might have answers for them myself when I next look into it, I'm just writing them down):
- even though LP: #2055239 is/was annoying, it helped to chase down this snap bug, so in the end it was useful. Are we sure we want to hide error messages?
- why are errors from apt-news and esm-cache sent to /dev/null, but the ones from apt-esm-json-hook not?
- why a .txt extension, and not .log?
- the way this is being done, the .txt file will be overwritten each time. If there was an error message in in before, it will be gone.
- conversely, if we append to the file instead of overwriting to it, we should think about its size growing without bounds
- why in `/var/lib/ubuntu-advantage` and not in `/var/log/<something>`?
- don't we have to clean this .txt file up on package removal?

Reply:
I checked some other scripts in /etc/apt/apt.conf.d/. Some redirect everything to /dev/null, others just redirect stdout, others just stderr, none to a file.

Have you considered using a proper log? Perhaps even `| logger ...`? For both perhaps?

Reply:
Thanks for thinking through this! I have some responses, but ultimately we need to make a call on the tradeoffs.

> I'll take a closer look at this on Monday. The questions that come to mind at the moment are (and I might have answers for them myself when I next look into it, I'm just writing them down):
> 
>     * even though LP: #2055239 is/was annoying, it helped to chase down this snap bug, so in the end it was useful. Are we sure we want to hide error messages?

*systemd (not snap)
I'd argue that it was only useful/necessary to fix LP: #2055239 because the messages from this hook were so annoying. IMO it's not that harmful a bug except for causing these messages that were very visible and made us look bad.

> 
>     * why are errors from apt-news and esm-cache sent to /dev/null, but the ones from apt-esm-json-hook not?
> 

The thing I had in mind with these redirections is: We still want to get this information if someone runs `pro collect-logs`. `pro collect-logs` already collects the status and logs of those systemd services - warnings/errors should be present there. The new log-like file was introduced so that we could collect apt-esm-json-hook info in `pro collect-logs`.

>     * why a .txt extension, and not .log?

As currently implemented it will only ever contain the most recent `stderr` output, so it's not really a log per se IMO, but I could be convinced otherwise.

> 
>     * the way this is being done, the .txt file will be overwritten each time. If there was an error message in in before, it will be gone.

Yes, I considered this a reasonable tradeoff for the simplicity of the approach. I worry about adding additional complexity and therefore ways that these hooks could fail.

> 
>     * conversely, if we append to the file instead of overwriting to it, we should think about its size growing without bounds

Yes, this is some of the complexity I avoided/punted on with the current approach

> 
>     * why in `/var/lib/ubuntu-advantage` and not in `/var/log/<something>`?

Same reason for not giving it a `.log` extension, but I could be convinced either way.

> 
>     * don't we have to clean this .txt file up on package removal?

Good point. So far we haven't removed anything in /var/lib/ubuntu-advantage on removal (but we do remove it entirely on purge). We could start removing this on removal though.

> I checked some other scripts in /etc/apt/apt.conf.d/. Some redirect everything to /dev/null, others just redirect stdout, others just stderr, none to a file.
> 
> Have you considered using a proper log? Perhaps even `| logger ...`? For both perhaps?

TIL about `logger` (thanks!). I'm only worried about unnecessary complexity and unnecessary visibility of errors. These aren't exactly critical services, so extra warnings/errors in syslog may be bothersome without much benefit. I could be convinced otherwise though. Do you think this would be useful and worth the tradeoffs?

Reply:
This whole approach sounds like it could use a second pass. As stated above, the moment I saw this, all these different questions popped into my mind. It just feels weird. If we are interested in possible errors messages, they should be logged. Writing them to some mysterious file in /var/lib is unexpected, and we will really only get the output from the last run, so I question its usefulness. Maybe start with a list:
* What are the kind of error messages we expect from each service?
* Is it bad if users see these messages? Why?
* Is it bad if the users DO NOT see these messages?
* apt-news, esm-cache, already log to ubuntu-advantage.log. Maybe the apt-es-json-hook should too
* I understand that the errors from apt-news did not come from apt-news per se, but from systemd/systemctl
* I also understand the extra complexity in getting apt-esm-json-hook to natively log somewhere (c++, hey). `logger` could help, but it's still extra complexity and not something to be rushed in at the eleventh hour.
* What are the consequences of NOT doing these changes at this time? Just more annoyance at that never-ending systemd/snap bug?
*

Reply:
Just to close this out. We discussed this out of band and decided on the fix implemented in #3288